### PR TITLE
Add link to protein id in the transcript accordion

### PR DESCRIPTION
**Pull request by msmanoj**
## Type

- [ ] Bug fix
- [ ] New feature
- [x] Improvement to existing feature

## Related JIRA Issue(s)
https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-723

## Description
Clicking on the protein id in the transcript accordion will now open the protein view, automatically expands the selected protein and scroll it into the view.
- Added loaders to both ProteinList & ProteinListItemInfo
- Added the functionality to expand the first transcript by default in both transcript and protein view
- Added the functionality to retain the scroll position when revisiting the transcript and protein view.
- Created `useRestoreScrollPosition` hook

## Deployment URL
http://link-protein-id.review.ensembl.org/entity-viewer/homo_sapiens_GCA_000001405_28/gene:ENSG00000139618?view=transcripts

## Views affected
Entity viewer -> Transcripts, Proteins

### Other effects
- [ ] I have checked any requirements for Google Analytics
- [ ] I have created any required tests
- [ ] The PR has as its final commit a WASM/JS update commit if any Rust files were updated.

### CODE REVIEW

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/transcripts-list-item-info/TranscriptsListItemInfo.tsx
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/transcripts-list-item-info/TranscriptsListItemInfo.tsx:143]: Looks like we could just use the `Link` component?

Comment (msmanoj) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/transcripts-list-item-info/TranscriptsListItemInfo.tsx:143]: >Regarding the expandProtein part, do we possibly want to add the focus protein to the url?

I thought the same but I am not sure when we need to update the focus protein. Do we do it on mouseover?  
Also, if we don't update it and if the user revisits the proteins list, the focus will always be the same.

Comment (azangru) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/transcripts-list-item-info/TranscriptsListItemInfo.tsx:143]: We may set just a unidirectional logic:

- if there is a focus-protein parameter in the url when the view is protein,
- then expand corresponding protein (if it exists)
- and scroll to the expanded protein

Not sure what should happen when user clicks on another protein. They should be allowed to expand multiple proteins.

Hmm... Maybe having focus protein id in the url is not such a good idea after all.

Comment (msmanoj) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/transcripts-list-item-info/TranscriptsListItemInfo.tsx:143]: I did think about this approach of having the focus parameter in the urls but I was wondering if we could simply scroll to the id that was most recently added to `expandedTranscriptIds`. With this approach, when the user clicks on a protein id, it will get added to the `expandedTranscriptIds` and will become the most recently added id. We will of course need to handle clicking on an already expanded id in which case we will have to pop it and add to the end.

Comment (azangru) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/transcripts-list-item-info/TranscriptsListItemInfo.tsx:143]: > if we could simply scroll to the id that was most recently added to expandedTranscriptIds

There are potential problems with this approach:
- I am not sure we even need to apply programmatic scrolling when the user is already on the protein page and is clicking on proteins to expand or collapse them. In this case, the user already sees the protein he is interacting with.
- How are we planning to handle page refreshes? Also, do we plan to be able to share the url with a particular expanded protein, or do we give up on this idea?

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/ProteinsList.scss
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/ProteinsList.scss:8]: Given this:

image

what's the purpose of height: 100% and flex-direction: column?

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/transcripts-list-item-info/TranscriptsListItemInfo.tsx
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/transcripts-list-item-info/TranscriptsListItemInfo.tsx:129]: the function says "buildUrl", but what it's building is a link element.

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/ProteinsList.scss
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/ProteinsList.scss:4]: why?

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item-info/ProteinsListItemInfo.scss
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item-info/ProteinsListItemInfo.scss:42]: Given this:

image

what's the purpose of height: 100% and flex-direction: column?

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item/ProteinsListItem.tsx
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item/ProteinsListItem.tsx:73]: "start" is the default

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item/ProteinsListItem.tsx
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item/ProteinsListItem.tsx:70]: I tried running it without the timeout, and it worked fine. At the point when `ProteinsListItem` has mounted, other `ProteinsListItem`s should exist as well. Is the timeout really necessary for anything?

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item/ProteinsListItem.scss
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item/ProteinsListItem.scss:24]: CSS rule `scroll-margin: 55px;` (or `scroll-margin-top: 55px`) applied to the ProteinsListItem div should make this span unnecessary. It currently has problems on Safari, but Safari will catch up sooner or later.

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/ProteinsList.scss
Comment (msmanoj) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/ProteinsList.scss]: I copied it from here:
https://github.com/Ensembl/ensembl-client/blob/dev/src/ensembl/src/content/app/entity-viewer/gene-view/GeneView.scss#L5

I have removed height and flex-direction as they don't make any difference.

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/transcripts-list-item-info/TranscriptsListItemInfo.tsx
Comment (msmanoj) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/transcripts-list-item-info/TranscriptsListItemInfo.tsx]: Renamed it to `getLinkToProteinView`

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/ProteinsList.scss
Comment (msmanoj) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/ProteinsList.scss]: I was playing around with overflow before the layout update PR got merged and I missed to clean this up.

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item-info/ProteinsListItemInfo.scss
Comment (msmanoj) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item-info/ProteinsListItemInfo.scss]: Removed them

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item/ProteinsListItem.tsx
Comment (msmanoj) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item/ProteinsListItem.tsx]: Removed it

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item/ProteinsListItem.tsx
Comment (msmanoj) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item/ProteinsListItem.tsx]: This was causing some issue when the newly focussed protein needs to expand along with the proteins that were already expanded previously. Now that we expand only the focussed protein, it will not be an issue anymore. I'll get it removed.

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item/ProteinsListItem.scss
Comment (msmanoj) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item/ProteinsListItem.scss]: That's done

**Comment by azangru** [comment]
Note from code review:
- return the setTimeout (prob 100ms?) before scrolling the transcript into view
- `transcriptId` -> `transcript_id` in url query

File: src/ensembl/src/content/app/entity-viewer/gene-view/GeneView.tsx
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/GeneView.tsx:194]: shouldn't need to create `new URLSearchParams(search)` twice.  `urlSearchParams = new URLSearchParams(search); const view = urlSearchParams.get('view'); const transcriptId = urlSearchParams.get('transcript_id)`

Comment (msmanoj) [reply src/ensembl/src/content/app/entity-viewer/gene-view/GeneView.tsx]: That's done

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/transcripts-list-item-info/TranscriptsListItemInfo.tsx
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/transcripts-list-item-info/TranscriptsListItemInfo.tsx:86]: `stableId` => `proteinStableId` (just to avoid confusion)?

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/transcripts-list-item-info/TranscriptsListItemInfo.tsx
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/transcripts-list-item-info/TranscriptsListItemInfo.tsx:91]: So, here, for example, I am already confused. Why is the `transcriptId` field receiving the protein id?

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item/ProteinsListItem.tsx
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item/ProteinsListItem.tsx:78]: Here, again, something that the code calls transcript id is compared to protein id.

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/transcripts-list-item-info/TranscriptsListItemInfo.tsx
Comment (msmanoj) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/transcripts-list-item-info/TranscriptsListItemInfo.tsx]: That's done

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/transcripts-list-item-info/TranscriptsListItemInfo.tsx
Comment (msmanoj) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/transcripts-list-item-info/TranscriptsListItemInfo.tsx]: Here and below, I have changed it all to proteinId. I was planning to update it today as it was getting late yesterday.

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/transcripts-list-item-info/TranscriptsListItemInfo.tsx
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/transcripts-list-item-info/TranscriptsListItemInfo.tsx:53]: `onProteinLinkClick` perhaps? It just looks so wrong that TranscriptListItemInfo should know something about proteins that are expanded in a completely different view.

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/ProteinsList.tsx
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/ProteinsList.tsx:69]: According to line 49 above (`proteinCodingTranscripts = sortedTranscripts.filter ...`), `proteinCodingTranscripts` are guaranteed to be at very least an empty array, which makes them always truthy; so the CircleLoader below isn't going to be rendered.

**Comment by azangru** [comment]
@msmanoj I pushed a commit that updated the function name (I didn't mean to rename the redux action, sorry for the confusion), fixed type errors (there were some) and added a test. Please let me know if you are happy with it.

**Comment by msmanoj** [comment]
> I didn't mean to rename the redux action, sorry for the confusion

Oops! sorry, it's just me being silly there by changing the redux action. Thanks for putting it back. The changes looks good 👍

**Comment by azangru** [comment]
@msmanoj I am seeing both vertical and horizontal scroll when I navigate from transcripts to proteins. is this intentional?

scroll

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item-info/ProteinsListItemInfo.tsx
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item-info/ProteinsListItemInfo.tsx:84]: `fetchProteinSummary` function now expects protein id. Changed in this PR: https://github.com/Ensembl/ensembl-client/commit/29b8f1cd5e83ac0bf63879b8d2ce3f1f86809bd7. Please check how the id should be passed to this function.

Comment (msmanoj) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item-info/ProteinsListItemInfo.tsx]: That's done

**Comment by msmanoj** [comment]
> @msmanoj I am seeing both vertical and horizontal scroll when I navigate from transcripts to proteins. is this intentional?
> 
> scroll

No, it's not intentional. `scrollIntoView` seems to be doing it automatically. I've fixed it now.

**Comment by msmanoj** [comment]
> We have a problem: `fetchProteinSummary` function can return null (it is actually returning null in my dev environment, ~I wonder why~ I left a comment explaining why). In this case, the spinner will remain visible indefinitely. We either need to hide it if the response is empty, or show the error/warning message. Don't know how much of this logic belongs in this PR, but eternal spinner is probably bad.

I've put a temporary fix for this now. Will look into it further when I work on https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-727.

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item-info/ProteinsListItemInfo.tsx
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item-info/ProteinsListItemInfo.tsx:115]: As a future todo: we have a lot of those magic `695`s floating around, both in typescript files and in scss files. Would be great if we could just use a single constant importable both in the script and the style files, like we did with colours for our palette.

Doesn't block this PR, but we should clean this up.

File: src/ensembl/src/shared/hooks/useRestoreScrollPosition.ts
Comment (azangru) [inline src/ensembl/src/shared/hooks/useRestoreScrollPosition.ts:42]: Is this comment still relevant?

**Review by azangru — APPROVED** [review]

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item-info/ProteinsListItemInfo.tsx
Comment (msmanoj) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/proteins-list/proteins-list-item-info/ProteinsListItemInfo.tsx:115]: Created a ticket to clean it up:
https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-800

**Pull request merged**
